From: Michal Vasko <mvasko@cesnet.cz>
Date: Thu, 14 Nov 2024 09:59:33 +0100
Subject: shm sub UPDATE optional full diff print/parse

It was done by default, now only the changes
of the specific module are sent.

(cherry picked from commit faf0fc0aa99b84dc39745e3ccddab886dba0cc0b)
---
 src/shm_sub.c       | 65 +++++++++++++++++++++++++++++++++++++++++++++--------
 src/sysrepo.c       |  3 ++-
 src/sysrepo_types.h |  8 ++++++-
 3 files changed, 65 insertions(+), 11 deletions(-)

diff --git a/src/shm_sub.c b/src/shm_sub.c
index 64a9fb1..753a7f8 100644
--- a/src/shm_sub.c
+++ b/src/shm_sub.c
@@ -1750,16 +1750,59 @@ sr_shmsub_change_notify_nsubs_set_mod_prio(struct sr_shmsub_many_info_change_s *
     *max_mpriority = cur_mprio - 1;
 }
 
+static sr_error_info_t *
+sr_shmsub_change_notify_get_diff(struct lyd_node *diff, const struct lys_module *ly_mod, int sub_opts,
+        char **full_diff_lyb, uint32_t *full_diff_lyb_len, char **diff_lyb, uint32_t *diff_lyb_len, int *free_diff)
+{
+    sr_error_info_t *err_info = NULL;
+    struct lyd_node *mod_diff;
+
+    /* free previous diff */
+    if (*free_diff) {
+        free(*diff_lyb);
+        *diff_lyb = NULL;
+    }
+
+    if (sub_opts & SR_SUBSCR_CHANGE_ALL_MODULES) {
+        /* print the full diff if not before */
+        if (!*full_diff_lyb && (err_info = sr_lyd_print_data(diff, LYD_LYB, 0, -1, full_diff_lyb, full_diff_lyb_len))) {
+            goto cleanup;
+        }
+
+        *diff_lyb = *full_diff_lyb;
+        *diff_lyb_len = *full_diff_lyb_len;
+        *free_diff = 0;
+    } else {
+        /* separate the diff of this module */
+        mod_diff = sr_module_data_unlink(&diff, ly_mod);
+        assert(mod_diff);
+
+        /* print it */
+        if ((err_info = sr_lyd_print_data(mod_diff, LYD_LYB, 0, -1, diff_lyb, diff_lyb_len))) {
+            goto cleanup;
+        }
+        *free_diff = 1;
+
+        /* relink to the diff */
+        if ((err_info = sr_lyd_insert_sibling(diff, mod_diff, &diff))) {
+            goto cleanup;
+        }
+    }
+
+cleanup:
+    return err_info;
+}
+
 sr_error_info_t *
 sr_shmsub_change_notify_change(struct sr_mod_info_s *mod_info, const char *orig_name, const void *orig_data,
         uint32_t timeout_ms, int update_flag, sr_error_info_t **cb_err_info)
 {
     sr_error_info_t *err_info = NULL;
-    uint32_t notify_count = 0, max_priority, cur_mpriority, diff_lyb_len, *aux = NULL, i, subscriber_count;
+    uint32_t notify_count = 0, max_priority, cur_mpriority, full_diff_lyb_len, diff_lyb_len, *aux = NULL, i, subscriber_count;
     struct sr_shmsub_many_info_change_s *notify_subs = NULL, *nsub;
     struct sr_mod_info_mod_s *mod = NULL;
-    char *diff_lyb = NULL;
-    int opts, pending_events;
+    char *full_diff_lyb = NULL, *diff_lyb = NULL;
+    int opts, pending_events, free_diff = 0;
     sr_cid_t cid;
     sr_sub_shm_t *sub_shm = NULL;
 
@@ -1814,11 +1857,6 @@ sr_shmsub_change_notify_change(struct sr_mod_info_s *mod_info, const char *orig_
     /* assign consolidated module priorities */
     sr_shmsub_change_notify_nsubs_set_mod_prio(notify_subs, notify_count, mod_info->ds, &cur_mpriority);
 
-    /* prepare the diff to write into subscription SHM */
-    if ((err_info = sr_lyd_print_data(mod_info->diff, LYD_LYB, 0, -1, &diff_lyb, &diff_lyb_len))) {
-        goto cleanup;
-    }
-
     do {
         pending_events = 0;
         for (i = 0; i < notify_count; ++i) {
@@ -1839,6 +1877,12 @@ sr_shmsub_change_notify_change(struct sr_mod_info_s *mod_info, const char *orig_
                 continue;
             }
 
+            /* prepare the all-modules diff to write into subscription SHM */
+            if ((err_info = sr_shmsub_change_notify_get_diff(mod_info->diff, nsub->mod->ly_mod, opts, &full_diff_lyb,
+                    &full_diff_lyb_len, &diff_lyb, &diff_lyb_len, &free_diff))) {
+                goto cleanup;
+            }
+
             /* open sub SHM and map it */
             if ((err_info = sr_shmsub_open_map(nsub->mod->ly_mod->name, sr_ds2str(mod_info->ds), -1, &nsub->shm_sub))) {
                 goto cleanup;
@@ -1948,7 +1992,10 @@ cleanup:
     }
 
     free(aux);
-    free(diff_lyb);
+    free(full_diff_lyb);
+    if (free_diff) {
+        free(diff_lyb);
+    }
     free(notify_subs);
     return err_info;
 }
diff --git a/src/sysrepo.c b/src/sysrepo.c
index 7510985..9399cf2 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -5573,7 +5573,8 @@ sr_module_change_subscribe(sr_session_ctx_t *session, const char *module_name, c
 
     conn = session->conn;
     /* only these options are relevant outside this function and will be stored */
-    sub_opts = opts & (SR_SUBSCR_DONE_ONLY | SR_SUBSCR_PASSIVE | SR_SUBSCR_UPDATE | SR_SUBSCR_FILTER_ORIG | SR_SUBSCR_PRODUCER_FILTERING);
+    sub_opts = opts & (SR_SUBSCR_DONE_ONLY | SR_SUBSCR_PASSIVE | SR_SUBSCR_UPDATE | SR_SUBSCR_FILTER_ORIG
+            | SR_SUBSCR_PRODUCER_FILTERING | SR_SUBSCR_CHANGE_ALL_MODULES);
 
     /* CONTEXT LOCK */
     if ((err_info = sr_lycc_lock(conn, SR_LOCK_READ, 0, __func__))) {
diff --git a/src/sysrepo_types.h b/src/sysrepo_types.h
index 2d4ccc4..74b9d12 100644
--- a/src/sysrepo_types.h
+++ b/src/sysrepo_types.h
@@ -464,6 +464,13 @@ typedef enum sr_subscr_flag_e {
      */
     SR_SUBSCR_FILTER_ORIG = 0x100000,
 
+    /**
+     * @brief If you call getters (::sr_get_data()) in the module change callback for data from **other** modules
+     * than the one the callback subscribed for, use this flag. Normally, only the changes of the subscribed module
+     * are sent to the callback so it retrieves the old data of other modules.
+     */
+    SR_SUBSCR_CHANGE_ALL_MODULES = 0x20000,
+
     /*
      * @brief For a given "get" operation on the operational datastore, this option ensures that a subscriber
      * is called no more than once, regardless of whether the subscription path is a child of a list.
@@ -476,7 +483,6 @@ typedef enum sr_subscr_flag_e {
      * this option is used so that filtering happens at the producer, instead of the event handler thread
      */
     SR_SUBSCR_PRODUCER_FILTERING = 0x200,
-
 } sr_subscr_flag_t;
 
 /**
